Use mempool only in progs linked to libsystemd-shared.so

keszybz opened the pull request.

yuwata commented on src/shared/meson.build:
```diff
@@ -36,6 +36,7 @@ shared_sources = files('''
         dropin.h
         efivars.c
         efivars.h
+        enable-mempool.c
```
??? enable-mempool.c is missing.

fsateler replied:
Wouldn't this include the constructor with the users of the static version? AFAICT that is not the desired result.

poettering commented:
How is the effect different from #10286? If a program links against both libsystemd-shared and against libsystemd (which is a bug btw, except if this happens due to NSS modules or PAM modules being loaded into systemd supplied programs) then they'll get a duplicate set of both the code and the internal variables of it, and they will not interfere with each other. This means that both with #10286 and with this PR you will have one set of hashmap functions that use the pool, and one set that does not, and depending which code calls either they'll get one effect or the other...

i mean, they key of the issue here is that the gcc constructor function would only initialize the copy in the private lib, but leave the one in the public lib as it is...

Or what am I missing?

poettering commented:
> Yeah, I don't think the _effect_ of my solution is better. What I wrote in a comment in #10286 was wrong. But my approach has an easier _implementation_ because it does not require two separate compilation objects. The build system is too complicated already.

i am not a fan of using gcc constructors btw... maybe another option is to simply define two .c files that define the global variable, and either one is compiled in or the other.  if both are accidentally pulled in there would be a link error, and if neither is pulled in too. But we'd nicely avoid the gcc constructors, i.e. pay no penalty at startup and avoid ugly ELF magic...

keszybz commented:
Updated along the requested lines. While looking at the meson files, I found something that I was pretty sure I had cleaned up before, and it turns out there was a branch I forgot to make a PR for. I included those commits here too.

yuwata commented:
Hmm. This does not fix the issue reported in https://github.com/systemd/systemd/pull/10021.

yuwata commented:
Moreover, this breaks `-Dstatic-libsystemd=true` or `-Dstatic-libudev=true`.
